packetbeat/protos/dns/dns_tcp.go

```diff
@@ -0,0 +1,302 @@
+	tcpTuple    *common.TcpTuple
+	rawData     []byte
+	parseOffset int
+	message     *DnsMessage
+}
+
+// dnsConnectionData contains two DnsStream's that hold data from a complete TCP
+// transmission. Element zero contains the response data. Element one contains
+// the request data.
+// prevRequest (previous Request) is used to add Notes to a transaction when a failing answer is encountered
+type dnsConnectionData struct {
+	Data        [2]*DnsStream
+	prevRequest *DnsMessage
+}
+
+func (dns *Dns) Parse(pkt *protos.Packet, tcpTuple *common.TcpTuple, dir uint8, private protos.ProtocolData) protos.ProtocolData {
+	defer logp.Recover("Dns ParseTcp")
+
+	logp.Debug("dns", "Parsing packet addressed with %s of length %d.",
+		pkt.Tuple.String(), len(pkt.Payload))
+
+	conn := ensureDnsConnection(private)
+
+	conn = dns.doParse(conn, pkt, tcpTuple, dir)
+	if conn == nil {
+		return nil
+	}
+
+	return conn
+}
+
+func ensureDnsConnection(private protos.ProtocolData) *dnsConnectionData {
+	if private == nil {
+		return &dnsConnectionData{}
+	}
+
+	conn, ok := private.(*dnsConnectionData)
+	if !ok {
+		logp.Warn("Dns connection data type error, create new one")
+		return &dnsConnectionData{}
+	}
+	if conn == nil {
+		logp.Warn("Unexpected: dns connection data not set, create new one")
+		return &dnsConnectionData{}
+	}
+
+	return conn
+}
+
+func (dns *Dns) doParse(conn *dnsConnectionData, pkt *protos.Packet, tcpTuple *common.TcpTuple, dir uint8) *dnsConnectionData {
+	stream := conn.Data[dir]
+	payload := pkt.Payload
+
+	if stream == nil {
+		stream = newStream(pkt, tcpTuple)
+	} else {
+		if stream.message == nil { // nth message of the same stream
+			stream.message = &DnsMessage{Ts: pkt.Ts, Tuple: pkt.Tuple}
+		}
+
+		stream.rawData = append(stream.rawData, payload...)
+		if len(stream.rawData) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("dns", "Stream data too large, dropping DNS stream")
+			conn.Data[dir] = nil
+			return conn
+		}
+	}
+	conn.Data[dir] = stream
+	decodedData, err := conn.Data[dir].handleTcpRawData()
+
+	if err != nil {
+		logp.Debug("dns", err.Error()+" addresses %s, length %d",
+			tcpTuple.String(), len(stream.rawData))
+
+		if err == IncompleteMsg {
+			logp.Debug("dns", "Waiting for more raw data")
+			return conn
+		}
+
+		if dir == tcp.TcpDirectionReverse {
+			dns.publishResponseError(conn, err)
+		}
+
+		// This means that malformed requests or responses are being sent...
+		// TODO: publish the situation also if Request
+		conn.Data[dir] = nil
+		return conn
+	}
+
+	dns.messageComplete(conn, tcpTuple, dir, decodedData)
+	conn.Data[dir].PrepareForNewMessage()
+	return conn
+}
+
+func newStream(pkt *protos.Packet, tcpTuple *common.TcpTuple) *DnsStream {
+	return &DnsStream{
+		tcpTuple: tcpTuple,
+		rawData:  pkt.Payload,
+		message:  &DnsMessage{Ts: pkt.Ts, Tuple: pkt.Tuple},
+	}
+}
+
+func (dns *Dns) messageComplete(conn *dnsConnectionData, tcpTuple *common.TcpTuple, dir uint8, decodedData *layers.DNS) {
```

Comment:
It might be needed in the future implementation of EDNS  and DNSSEC, what's why the function wasn't removed. Please tell me if that's not appropriate.

Thank you for the review btw.